arch/cortex-m/src/syscall.rs

```diff
@@ -0,0 +1,187 @@
+        // called once.
+        write_volatile(&mut APP_FAULT, 0);
+
+        // Check to see if the svc_handler was called and the process called a
+        // syscall.
+        let syscall_fired = read_volatile(&SYSCALL_FIRED);
+        write_volatile(&mut SYSCALL_FIRED, 0);
+
+        if app_fault == 1 {
+            // APP_FAULT takes priority. This means we hit the hardfault handler
+            // and this process faulted.
+            kernel::syscall::ContextSwitchReason::Fault
+        } else if syscall_fired == 1 {
+            kernel::syscall::ContextSwitchReason::SyscallFired
+        } else {
+            kernel::syscall::ContextSwitchReason::Other
+        }
+    }
+
+    /// Get the syscall that the process called.
+    unsafe fn get_syscall(&self, stack_pointer: *const usize) -> Option<kernel::syscall::Syscall> {
+        // Get the four values that are passed with the syscall.
+        let r0 = read_volatile(stack_pointer.offset(0));
+        let r1 = read_volatile(stack_pointer.offset(1));
+        let r2 = read_volatile(stack_pointer.offset(2));
+        let r3 = read_volatile(stack_pointer.offset(3));
+
+        // Get the actual SVC number.
+        let pcptr = read_volatile((stack_pointer as *const *const u16).offset(6));
+        let svc_instr = read_volatile(pcptr.offset(-1));
+        let svc_num = (svc_instr & 0xff) as u8;
+        match svc_num {
+            0 => Some(kernel::syscall::Syscall::YIELD),
+            1 => Some(kernel::syscall::Syscall::SUBSCRIBE {
+                driver_number: r0,
+                subdriver_number: r1,
+                callback_ptr: r2 as *mut (),
+                appdata: r3,
+            }),
+            2 => Some(kernel::syscall::Syscall::COMMAND {
+                driver_number: r0,
+                subdriver_number: r1,
+                arg0: r2,
+                arg1: r3,
+            }),
+            3 => Some(kernel::syscall::Syscall::ALLOW {
+                driver_number: r0,
+                subdriver_number: r1,
+                allow_address: r2 as *mut u8,
+                allow_size: r3,
+            }),
+            4 => Some(kernel::syscall::Syscall::MEMOP {
+                operand: r0,
+                arg0: r1,
+            }),
+            _ => None,
+        }
+    }
+
+    unsafe fn set_syscall_return_value(&self, stack_pointer: *const usize, return_value: isize) {
+        // For the Cortex-M arch we set this in the same place that r0 was
+        // passed.
+        let sp = stack_pointer as *mut isize;
+        write_volatile(sp, return_value);
+    }
+
+    unsafe fn pop_syscall_stack(
+        &self,
+        stack_pointer: *const usize,
+        state: &mut CortexMStoredState,
+    ) -> *mut usize {
+        state.yield_pc = read_volatile(stack_pointer.offset(6));
+        state.psr = read_volatile(stack_pointer.offset(7));
+        (stack_pointer as *mut usize).offset(8)
+    }
+
+    unsafe fn push_function_call(
```

Comment:
On the same line as Phil's comment about pop, what if `push_function_call` was called when the process's stack is entirely full? Is there a way we can check that we're still within the bounds of process memory before writing to the stack pointer? If we aren't, then the response can be to put the process into a fault state.